### ISSUE TITLE
VAT is not automaticly added to invoices - wrong price results! [V7,V8,Master]

### ISSUE THREAD

**Issue by Yenthe666** [odoo/odoo#5117]
**VAT is ignored - not automaticly calculated**

Impacted versions:

 - 7.0, 8.0 and master

Steps to reproduce:

 1. Create a new invoice
 2. Add a product with a tax (for example a product with price 100.00 and tax VAT-OUT-21-S (21%))
 3. Click on 'Save'

Current behavior:

 - Total price of 100.0, the taxes are not automaticly added.

Expected behavior:

 - Total price of 121.00 (100 without tax + 21% tax = 121)<br />
Quotations and sales orders automaticly add and calculate the taxes, invoices do not. This should have the same logic because this is not logical and its easy to miss!
Why isn't this logic here for invoices too?

Sample without clicking on update:
<img src="http://i.imgur.com/sPey8lq.png"/>

Sample after clicking on update:
<img src="http://i.imgur.com/L1Pd66J.png"/>

**Comment by Yenthe666** [comment]
@xmo-odoo could you have a look at this? This is wrong in all versions and seems like a major flaw in my idea..

**Comment by Yenthe666** [comment]
@odony this might be something for you too.

**Comment by Yenthe666** [comment]
<b>Update:</b> The tax seems to be added when you validate an invoice. But this should happen <b>before</b> validating it, when saving the invoice.
@odony I'd like a statement on this please. @xmo-odoo interesting for you too.

**Comment by odony** [comment]
tl;dr: This is the intended behavior.

Long answer:
The taxes *are* correctly added on the invoice lines. What is not done at the same time is the computation of the **tax lines** and the corresponding subtotal.

The process for computing taxes on invoices is quite different from what happens on a SO or PO. On a SO/PO the system simply computes the taxes subtotal along with the other amounts, whenever the SO/PO is modified.

For invoices the system has to generate the whole list of tax lines, grouping the taxes from all the different product lines and computing the base amount and tax amount for each individual tax. 
This can be an expensive operation so it should not be done every time you touch a line or save the invoice. Imagine working on invoices with hundreds of lines, you don't want the system to freeze for the recomputation every time you save, especially if the invoice is not final yet.
That's why there is an "update" button next to the Tax subtotal while you are editing an invoice, in order to update the taxes manually. And the system does a final recomputation of the taxes when the invoice is validated, to make sure the tax lines are up to date.
